synapse/storage/databases/main/cache.py

```diff
@@ -483,6 +483,30 @@ def _invalidate_cache_and_stream(
         txn.call_after(cache_func.invalidate, keys)
         self._send_invalidation_to_replication(txn, cache_func.__name__, keys)
 
+    def _invalidate_cache_and_stream_bulk(
+        self,
+        txn: LoggingTransaction,
+        cache_func: CachedFunction,
+        key_tuples: Collection[Tuple[Any, ...]],
+    ) -> None:
+        """A bulk version of _invalidate_cache_and_stream.
+
+        Locally invalidate every key-tuple in `key_tuples`, then emit invalidations
+        for each key-tuple over replication.
+
+        This implementation is more efficient than a loop which repeatedly calls the
+        non-bulk version.
+        """
+        if not key_tuples:
+            return
+
+        for keys in key_tuples:
+            txn.call_after(cache_func.invalidate, keys)
```

Comment:
@DMRobertson postulated if it would be better to do:

```suggestion
        def invalidate():
            for keys in key_tuples:
                cache_func.invalidate(keys)
        txn.call_after(invalidate)
```

But I don't think it matters much since `txn_call_after` just calls them in a loop anyway? Maybe one way would use less memory. 🤷